File: src/scraping.clj
 (ns scraping)

 (def base-url "http://www.pythonchallenge.com/pc/def/linkedlist.php?nothing=")

 (defn get-page [num]
   (slurp (str base-url num)))

 (defn get-num [page]
   (get (re-find #"nothing is (\d+)" page) 1))


 (defn find-last
   ([last-num] (find-last last-num []))
   ([last-num results]
    (let [num (get-num (get-page last-num))]
      (if (nil? num)
        results
        (find-last num (conj results num))))))

;(println (find-last 12345))
; divided by two
(println (find-last 8022))

File: src/scraping.hs
import Network.HTTP
import Text.Regex

baseUrl = "http://www.pythonchallenge.com/pc/def/linkedlist.php?nothing="

openURL :: String -> IO String
openURL x = getResponseBody =<< simpleHTTP (getRequest x)

getNum :: String -> Maybe Int
getNum page = do
    match <- matchRegex regex page
    return $ read . head $ match
    where
        regex = mkRegex "nothing is ([0-9]+)"

getPage :: Int -> IO String
getPage num = openURL $ baseUrl ++ (show num)

getNewNum :: Int -> IO (Maybe Int)
getNewNum num = do
    page <- getPage num
    return $ getNum page

findLast :: Int -> IO Int
findLast initNum = do
    newNum <- getNewNum initNum
    case newNum of
        Nothing -> return $ initNum
        (Just n) -> findLast n

main :: IO ()
main = do
    num <- findLast 8022
    print num
    return ()
